snapcraft/internal/project_loader/_config.py

```diff
@@ -294,3 +395,50 @@ def _expand_filesets_for(step, properties):
             new_step_set.append(item)
 
     return new_step_set
+
+
+class _Architecture:
+    def __init__(self, *, build_on, run_on=None):
+        self.build_on = build_on
+        if isinstance(build_on, str):
+            self.build_on = [build_on]
+
+        self.run_on = run_on
+        if isinstance(run_on, str):
+            self.run_on = [run_on]
+
+        # If there is no run_on, it defaults to the value of build_on
+        if not run_on:
+            self.run_on = self.build_on
+
+
+def _get_architecture_list(architectures, current_arch):
+    if not architectures:
+        return [_Architecture(build_on=[current_arch])]
+
+    build_architectures = []  # type: List[str]
+    architecture_list = []  # type: List[_Architecture]
+    for item in architectures:
+        if isinstance(item, str):
+            build_architectures.append(item)
+        if isinstance(item, dict):
+            architecture_list.append(_Architecture(
+                build_on=item.get('build-on'),
+                run_on=item.get('run-on')))
+
+    if build_architectures:
+        architecture_list.append(_Architecture(build_on=build_architectures))
+
+    return architecture_list
+
+
+def _process_architectures(architectures, project):
```

Comment:
why not just pass in `deb_arch`?

Reply:
Yeah that works :stuck_out_tongue: . Done!